src/plugins/data/server/saved_objects/search_migrations.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { flow, get } from 'lodash';
+import { SavedObjectMigrationFn } from 'kibana/server';
+
+const migrateIndexPattern: SavedObjectMigrationFn = doc => {
+  const searchSourceJSON = get(doc, 'attributes.kibanaSavedObjectMeta.searchSourceJSON');
+  if (typeof searchSourceJSON !== 'string') {
+    return doc;
+  }
+  let searchSource;
+  try {
+    searchSource = JSON.parse(searchSourceJSON);
+  } catch (e) {
+    // Let it go, the data is invalid and we'll leave it as is
+    return doc;
+  }
+
+  if (searchSource.index && Array.isArray(doc.references)) {
```

Comment:
I am curious why we decided to protect against `doc.references` not being an array this way. Should existing references not existing prevent us from migrating the `searchSource`?

Reply:
For reference, here is what we are migrating from:

https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/kibana/migrations/migrations.js#L38-L62

Reply:
@tylersmalley not sure that your case is possible cause in version 7.0.0 we added  `references` for all all saved object types. Code you can find in that file. 

But why it was added: 
Please open `src/core/server/saved_objects/serialization/types.ts` file and scroll to line: 
`export type SavedObjectUnsanitizedDoc = SavedObjectDoc & Partial<Referencable>;`

As you see we add `references` to `SavedObjectUnsanitizedDoc` using  `Partial` utility type.  

This means that if I remove the check in my migration script, I will encounter the following ts error:

![image](https://user-images.githubusercontent.com/20072247/79574196-7a2ff800-80c8-11ea-8d7b-6a6acd493fd3.png)

Reply:
I am not sure if there is an actual issue here, but I wanted to also bring it up so we can avoid this in the future. When we have a type issue like the one seen here where `Object is possibly 'undefined'`. We shouldn't change the behavior of the function in order to resolve the type error. In this case, if someone did call `migrateIndexPattern` where `doc.references` was not an array it would result in an undesired behavior and behavior that differed from the original implementation (with migrations it's also important that we don't change the behavior). In this case, I feel the correct resolution would be so set `doc.references` in the method (as done in `setNewReferences` before calling `migrateIndexPattern`, or casting the type to something which does not have references as optional.

Reply:
@tylersmalley I also thought about it today, and agree with you it should be fixed. I'll do it soon.